src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.DecryptTokenResult.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Text;
+using Microsoft.IdentityModel.Tokens;
+using Microsoft.IdentityModel.Logging;
+using Microsoft.IdentityModel.Abstractions;
+using TokenLogMessages = Microsoft.IdentityModel.Tokens.LogMessages;
+using System.Diagnostics;
+
+namespace Microsoft.IdentityModel.JsonWebTokens
+{
+    public partial class JwtTokenUtilities
+    {
+        /// <summary>
+        /// Decrypts a JWT token.
+        /// </summary>
+        /// <param name="jsonWebToken">The JWT token to decrypt.</param>
+        /// <param name="validationParameters">The <see cref="ValidationParameters"/> to be used for validating the token.</param>
+        /// <param name="decryptionParameters">The decryption parameters container.</param>
+        /// <param name="callContext">The call context used for logging.</param>
+        /// <returns>The decrypted, and if the 'zip' claim is set, decompressed string representation of the token.</returns>
+        internal static TokenDecryptingResult DecryptJwtToken(
+            JsonWebToken jsonWebToken,
+            ValidationParameters validationParameters,
+            JwtTokenDecryptionParameters decryptionParameters,
+            CallContext callContext)
+        {
+            if (validationParameters == null)
+                return new TokenDecryptingResult(
+                    jsonWebToken,
+                    ValidationFailureType.NullArgument,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10000,
+                            nameof(validationParameters)),
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            if (decryptionParameters == null)
+                return new TokenDecryptingResult(
+                    jsonWebToken,
+                    ValidationFailureType.NullArgument,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10000,
+                            nameof(decryptionParameters)),
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            bool decryptionSucceeded = false;
+            bool algorithmNotSupportedByCryptoProvider = false;
+            byte[] decryptedTokenBytes = null;
+
+            // keep track of exceptions thrown, keys that were tried
+            StringBuilder exceptionStrings = null;
+            StringBuilder keysAttempted = null;
+            string zipAlgorithm = null;
+            foreach (SecurityKey key in decryptionParameters.Keys)
```

Comment:
Consider refactoring and adding a method that takes one key.
This will simplify the logic.

Reply:
Adding a work item to look at this after we have our `ValidateTokenAsync` implementation working.